---
title: "satuRn: differential equivalence class usage analysis"
author: "Jeroen Gilis"
date: "08/07/2022"
output:
  html_document:
    code_download: yes
    theme: cosmo
    toc: yes
    toc_float: yes
    highlight: tango
    number_sections: yes
  pdf_document:
    toc: yes
---

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FALSE}
library(edgeR)
library(SummarizedExperiment)
library(satuRn)
library(ggplot2)
```

# Import TCC

Note, we provide the TCC matrix in the Extended data of our F1000 publication.
This matrix was generated by quantifying the scRNA-seq dataset by 
[Tasic et al.](https://doi.org/10.1038/s41586-018-0654-5) using Salmon v1.1.0
and setting the `--dumpEq` flag. This generates files containing quantitative
information on the equivalence classes. To import and wrangle these files
into the rds object `Tasic_TCC_mat.rds`, we used the `salmon-EC` from the
R package [fishpond](https://doi.org/doi:10.18129/B9.bioc.fishpond).

```{r}
TCC_mat <- readRDS("./Data/Tasic_TCC_mat.rds")
TCC_mat[1:5,1:5]
```

# Import transript-gene info

Obtained from `AnnotationHub`.

```{r}
tx2gene <- readRDS("./Data/Tasic_tx2gene.rds")
head(tx2gene)
```

# Import metadata

Metadata was obtained from the original publication by
[Tasic et al.](https://doi.org/10.1038/s41586-018-0654-5) and made available
from the Case_study.zip folder from our 
[Zenodo repository](https://doi.org/10.5281/zenodo.4439415).

```{r}
access <- read.csv2("./Data/Tasic_metadata_2.csv",sep = "\t")
access <- access[match(colnames(TCC_mat),access$SRA_Run),]

metaData <- openxlsx::read.xlsx("./Data/Tasic_metadata_1.xlsx")
metaData <- metaData[match(access$sample_name,metaData$sample_name),]
```

```{r}
all(colnames(TCC_mat) == access$SRA_Run)
all(access$sample_name == metaData$sample_name)

colnames(TCC_mat) <- metaData$sample_name
```

# Filtering

## Filtering on metadata

```{r}
# only retain cell that were unambiguously assigned to a certain cell type (cluster)
# note: could additionally filter on e.g. total_reads

metaData <- metaData[metaData$core_intermediate_call == "Core",]

metaData$group <- paste(metaData$brain_region,metaData$cluster,sep=".")
remove <- names(table(interaction(metaData$brain_region,metaData$cluster))[table(interaction(metaData$brain_region,metaData$cluster)) < 30])
metaData <- metaData[-which(metaData$group %in% remove),]

TCC_mat <- TCC_mat[,metaData$sample_name]
metaData <- metaData[,c("sample_name", "brain_region","cluster")]
```

## Filter ECs

Prior to filtering the count matrix using `edgeR::filterByExpr`, I will 
prefilter the data manually, requiring a expression of at least 10 counts when 
summing counts of all cells (extremely lenient pre-filtering which shouldn't 
change the end result, just to speed up the process).

```{r}
TCC_mat <- TCC_mat[-which(rowSums(TCC_mat)<10),]
nrow(TCC_mat)
```

Next, I remove ECs that are the only expressed EC in the corresponding gene:

```{r}
# Gene-level counts based on EC data
eccs <- strsplit(rownames(TCC_mat),"\\|",fixed=FALSE)
eccs <- lapply(eccs, function(x) x[1]) # we can simply take the first transcript
# identifier, since we already made sure all transcripts in the equivalence 
# class come from tehe same gene.
geneForEachEC <- as.factor(tx2gene[as.integer(unlist(eccs))+1,"gene_id"])
```

```{r}
# Remove EC that are the only EC expressed of a certain gene
uniqueGene <- which(isUnique(geneForEachEC))
geneForEachEC <- geneForEachEC[-uniqueGene]
TCC_mat <- TCC_mat[-uniqueGene,]
nrow(TCC_mat)
```

Finally, I run `edgeR::filterByExpr` using the same parameters as in the 
tx-level analysis.

```{r}
filter_edgeR <- filterByExpr(TCC_mat,
                             design = NULL,
                             group = metaData$brain_region,
                             lib.size = NULL,
                             min.count = 10,
                             min.total.count = 0,
                             large.n = 0,
                             min.prop = 0.7)
table(filter_edgeR)
TCC_mat <- TCC_mat[filter_edgeR,] # 28.462 ECs left
```

```{r}
# Update geneForEachECC after filtering
eccs <- strsplit(rownames(TCC_mat),"\\|",fixed=FALSE)
eccs <- lapply(eccs, function(x) x[1])
geneForEachEC <- as.factor(tx2gene[as.integer(unlist(eccs))+1,"gene_id"])

# Remove ECs that are the only EC expressed of a certain gene after filtering
uniqueGene <- which(isUnique(geneForEachEC))
```

```{r}
geneForEachEC <- geneForEachEC[-uniqueGene]
TCC_mat <- TCC_mat[-uniqueGene,]
dim(TCC_mat) # 26.141 ECs left
```

```{r}
eccs <- strsplit(rownames(TCC_mat),"\\|",fixed=FALSE)
eccs <- lapply(eccs, function(x) x[1])
geneForEachEC <- as.factor(tx2gene[as.integer(unlist(eccs))+1,"gene_id"])
```

# satuRn analysis

## Wrangle input

```{r, message=F}
eccInfo <- data.frame(isoform_id = rownames(TCC_mat),
                     gene_id = as.character(geneForEachEC))
sumExp <- SummarizedExperiment::SummarizedExperiment(assays=list(counts=TCC_mat), 
                                                     colData = metaData, 
                                                     rowData = eccInfo)
```

## Fit GLMs

```{r}
sumExp <- satuRn::fitDTU(object = sumExp,
                         formula = ~ 0 + cluster,
                         parallel = FALSE,
                         BPPARAM = BiocParallel::bpparam(),
                         verbose = TRUE)
```

## Contrast matrix

```{r}
# We here manually construct the contrast of interest as defined by Tasic et al.
design <- model.matrix(~0+sumExp$cluster)
colnames(design) <- levels(as.factor(sumExp$cluster))

L <- matrix(0, ncol = 8, nrow = ncol(design))
rownames(L) <- colnames(design)
colnames(L) <- c("C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")

L[c("L5 IT VISp Batf3", "L5 IT ALM Cpa6 Gpr88"), 1] <- c(1,-1)
L[c("L5 IT VISp Col27a1", "L5 IT ALM Cbln4 Fezf2"), 2] <- c(1,-1)
L[c("L5 IT VISp Col6a1 Fezf2", "L5 IT ALM Cpa6 Gpr88"), 3] <- c(1,-1)
L[c("L5 IT VISp Col6a1 Fezf2", "L5 IT ALM Gkn1 Pcdh19"), 4] <- c(1,-1)
L[c("L5 IT VISp Hsd11b1 Endou", "L5 IT ALM Lypd1 Gpr88"), 5] <- c(1,-1)
L[c("L5 IT VISp Hsd11b1 Endou", "L5 IT ALM Tnc"), 6] <- c(1,-1)
L[c("L5 IT VISp Hsd11b1 Endou", "L5 IT ALM Tmem163 Dmrtb1"), 7] <- c(1,-1)
L[c("L5 IT VISp Whrn Tox2", "L5 IT ALM Tmem163 Arhgap25"), 8] <- c(1,-1)
L
```

## Test contrasts

```{r}
sumExp <- satuRn::testDTU(object = sumExp,
                          contrasts = L,
                          diagplot1 = TRUE,
                          diagplot2 = TRUE,
                          sort = FALSE)
```

# Read transcript-level results

The results of the transcript-level `satuRn` analysis on the Tasic dataset can
be imported from the data folder. These results were generated in the script
**satuRn_analysis.Rmd**, which can be obtained from the **CaseStudy** folder.

```{r}
sumExp_tx <- readRDS(file = "./Data/Tasic_caseStudy_satuRn.Rds")
```

```{r}
rowData(sumExp_tx)$isoform_id[rowData(sumExp_tx)$gene_id == "ENSMUSG00000029470"]
rowData(sumExp)$isoform_id[rowData(sumExp)$gene_id == "ENSMUSG00000029470"]
```

# Publication Figure 11 and Extended data Figure 27

Here, we visualize the differential usage of features of two genes, 
ENSMUSG00000021699 and ENSMUSG00000029470, both for the TCC-level analysis and
the transcript-level analysis.

**These plots are used in the `satuRn` manuscript in Figure 11 and Figure S27.**

```{r}
interest <- c("ENSMUSG00000021699", "ENSMUSG00000029470")

for (i in seq_along(interest)) {

    # EC-level plot
    group1 <- colnames(sumExp)[sumExp$cluster == "L5 IT VISp Hsd11b1 Endou"]
    group2 <- colnames(sumExp)[sumExp$cluster == "L5 IT ALM Tnc"]
    
    plots <- satuRn::plotDTU(object = sumExp, 
                             contrast = "C6", 
                             groups = list(group1,group2), 
                             coefficients = list(c(rep(0,12),1,rep(0,1)),
                                                 c(rep(0,8),1,rep(0,5))), 
                             summaryStat = c("model","mean"), 
                             transcripts = NULL,
                             genes = interest[i],
                             top.n = 6)
    
    pdf(file = paste0("./Results/TCC_vs_tx/", interest[i], "_TCC.pdf"),
        width = 4.5,
        height = 4.5)
    
    if(i == 2){ # rearrange for manuscript
      plots <- plots[c(4,1,2,3)]
    }
    
    for (j in seq_along(plots)) {
        current_plot <- plots[[j]]  + 
            scale_fill_manual(labels = c("VISp","ALM"), values=c("royalblue4", "firebrick")) +
            scale_x_discrete(labels= c("Hsd11b1_Endou","Tnc")) +
            theme(axis.text.x = element_text(angle = 0, vjust = 1, hjust = 0.5, size = 9)) + 
            theme(strip.text = element_text(size = 9, face = "bold")) +
            ggtitle(paste0("EC ",j)) + 
            theme(plot.title = element_text(size=12,hjust = 0.5))
        print(current_plot)
    }
    
    dev.off()
    
    # transcript-level plot
    group1 <- colnames(sumExp_tx)[sumExp_tx$cluster == "VISp.L5_IT_VISp_Hsd11b1_Endou"]
    group2 <- colnames(sumExp_tx)[sumExp_tx$cluster == "ALM.L5_IT_ALM_Tnc"]

    plots <- satuRn::plotDTU(object = sumExp_tx,
                             contrast = "C6",
                             groups = list(group1,group2),
                             coefficients = list(c(rep(0,12),1,rep(0,1)),
                                                 c(rep(0,8),1,rep(0,5))),
                             summaryStat = c("model","mean"),
                             transcripts = NULL,
                             genes = interest[i],
                             top.n = 6)

    pdf(file = paste0("./Results/TCC_vs_tx/", interest[i], "_tx.pdf"),
        width = 4.5,
        height = 4.5)

    for (j in seq_along(plots)) {
        current_plot <- plots[[j]]  +
            scale_fill_manual(labels = c("VISp","ALM"), values=c("royalblue4", "firebrick")) +
            scale_x_discrete(labels= c("Hsd11b1_Endou","Tnc")) +
            theme(axis.text.x = element_text(angle = 0, vjust = 1, hjust = 0.5, size = 9)) +
            theme(strip.text = element_text(size = 9, face = "bold")) +
            ggtitle(rowData(sumExp_tx)$isoform_id[rowData(sumExp_tx)$gene_id == interest[i]][j]) + 
            theme(plot.title = element_text(size=10,hjust = 0.5))
        print(current_plot)
    }

    dev.off()
}
```

# Publication Table 2

Number of differentially used TCCs for the different contrasts

```{r}
sum(rowData(sumExp)[["fitDTUResult_C1"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C2"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C3"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C4"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C5"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C6"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C7"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp)[["fitDTUResult_C8"]]$empirical_FDR < 0.05, na.rm = T)
```

Number of unique genes with differentially used TCCs for the different contrasts

```{r}
genes_TCC1 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C1"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC2 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C2"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC3 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C3"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC4 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C4"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC5 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C5"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC6 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C6"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC7 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C7"]]$empirical_FDR < 0.05),
                                     "gene_id"])
genes_TCC8 <- unique(rowData(sumExp)[which(rowData(sumExp)[["fitDTUResult_C8"]]$empirical_FDR < 0.05),
                                     "gene_id"])
length(genes_TCC1)
length(genes_TCC2)
length(genes_TCC3)
length(genes_TCC4)
length(genes_TCC5)
length(genes_TCC6)
length(genes_TCC7)
length(genes_TCC8)
```

Number of differentially used transcripts for the different contrasts

```{r}
sum(rowData(sumExp_tx)[["fitDTUResult_C1"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C2"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C3"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C4"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C5"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C6"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C7"]]$empirical_FDR < 0.05, na.rm = T)
sum(rowData(sumExp_tx)[["fitDTUResult_C8"]]$empirical_FDR < 0.05, na.rm = T)
```

Number of unique genes with differentially used transcripts for the different 
contrasts

```{r}
genes_Tx1 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C1"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx2 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C2"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx3 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C3"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx4 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C4"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx5 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C5"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx6 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C6"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx7 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C7"]]$empirical_FDR < 0.05),
                                       "gene_id"])
genes_Tx8 <- unique(rowData(sumExp_tx)[which(rowData(sumExp_tx)[["fitDTUResult_C8"]]$empirical_FDR < 0.05),
                                       "gene_id"])
length(genes_Tx1)
length(genes_Tx2)
length(genes_Tx3)
length(genes_Tx4)
length(genes_Tx5)
length(genes_Tx6)
length(genes_Tx7)
length(genes_Tx8)
```

```{r}
length(intersect(genes_Tx1, genes_TCC1))
length(intersect(genes_Tx2, genes_TCC2))
length(intersect(genes_Tx3, genes_TCC3))
length(intersect(genes_Tx4, genes_TCC4))
length(intersect(genes_Tx5, genes_TCC5))
length(intersect(genes_Tx6, genes_TCC6))
length(intersect(genes_Tx7, genes_TCC7))
length(intersect(genes_Tx8, genes_TCC8))
```


