---
title: "1_Gtex_prepare"
author: "Jeroen Gilis"
date: "02/11/2020"
output: html_document
---

**In order to run this script (1_Gtex_perpare.Rmd), two files should be downloaded from Zenodo and copied into the Data folder from this Github page**. Note that if you want to run the analysis on scaledTPM expression values, you must also copy a third file from Zenodo into the Data folder from this Github page

1. GTEx_counts.gz 
2. 01_GTEx_baseline_metadata.txt
(3. GTEx_scaledTPM.gz)

Note that these files are identical to the ones from the links above; we have only renamed the files.

1. https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RSEMv1.3.0_transcript_expected_count.gct.gz
2. https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt
3. https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RSEMv1.3.0_transcript_tpm.gct.gz

Note that the raw data (e.g. .bam files) can also be acquired through [the GTEx data portal](https://www.gtexportal.org/home/documentationPage#staticTextPublicationPolicy)  

**If you do not want to run this script, its output can also be downloaded from Zenodo: GTEx_benchmark_datasets_count.Rdata (or, alternatively, GTEx_benchmark_datasets_scaledTPM.Rdata) **

Given the large datasets, this script is quite memory intensive!

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FALSE}
library(data.table)
library(edgeR)
library(DRIMSeq)
library(doMC)
```

# Load Gtex data

```{r,"load data"}
gtexCm <- fread("./Data/GTEx_counts.gz", data.table = FALSE)
gtexSample <- fread("./Data/GTEx_metadata.txt", data.table = FALSE)
```

# Subset Gtex data

The Gtex dataset originates from an observational study. To use it as a benchmark dataset, we need to know the ground truth, i.e. which transcripts are differentially used between groups of samples. To this end, we first generate mock datasets where no DTU is expected, after which we introduce DTU between groups by the transcript-abundance-swapping strategy discussed by Van den Berge et al. [stageR paper](https://doi.org/10.1186/s13059-017-1277-0).

In order to obtain a mock dataset, however, we need to construct a dataset where (ideally) no DTU is expected. We try to achieve this by subsampling a homogenous set of samples from the Gtex data object, based on following criteria:

- Only adrenal gland data 
- Only RNA Extraction from Paxgene-derived Lysate Plate Based samples
- Only samples from center B1
- Only RNASEQ samples in GTEx Analysis Freeze (SMAFRZE)

```{r}
# Extract gene info
txInfo <- gtexCm[,c('transcript_id','gene_id')]
colnames(txInfo)[1] <- 'isoform_id'
rownames(gtexCm) <- txInfo$isoform_id

# Extract samples
gtexSample <- gtexSample[
    which(gtexSample$SMTSD == 'Adrenal Gland'),
]

# Specifically filter on extraction kit
gtexSample <- gtexSample[
    which(gtexSample$SMNABTCHT == 'RNA Extraction from Paxgene-derived Lysate Plate Based'),
]

# Specifically filter on SMAFRZE
gtexSample <- gtexSample[
    which(gtexSample$SMAFRZE == 'RNASEQ'),
]

# Specifically filter on center
gtexSample <- gtexSample[
    which(gtexSample$SMCENTER == 'B1'),
]

gtexCm <- gtexCm[,which(
    colnames(gtexCm) %in% gtexSample$SAMPID
)]
```

Uncomment this chunk of code if scaledTPM expression values need to be obained instead of raw counts

```
colsums_count <- colSums(gtexCm)
colsums_count <- readRDS("./Data/colsums_count.Rds")
rm("gtexCm")
invisible(gc())
gtexTPM <- fread('./Data/GTEx_scaledTPM.gz', data.table = FALSE)

gtexTPM <- gtexTPM[,which(
    colnames(gtexTPM) %in% gtexSample$SAMPID
)]

# Get gtexScaledTPM 
colsums_TPM <- colSums(gtexTPM)
scale_factors <- colsums_count/colsums_TPM

gtexScaledTPM <- as.matrix(gtexTPM) %*% diag(scale_factors) # scaledTPM input
gtexScaledTPM <- as.data.frame(gtexScaledTPM)
colnames(gtexScaledTPM) <- colnames(gtexTPM)
rownames(gtexScaledTPM) <- txInfo$isoform_id
gtexCm <- gtexScaledTPM
```

# Setup generation of benchmark data

Set hyperparameters of the benchmark datasets (i.e. sample size, number of repeats, fraction of DTU genes and parameters for parallel processing).

```{r}
# Set benchmark parameters
samplesPrCondition   <- c(5,20,50) 
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

# Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'
    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

# list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
```

# Generate gtex benchmark data

```{r}
source(file="./Performance_benchmarks/getBenchmark_data.R")
```

```{r}
gtexBenchmarkDataLenient <- getBenchmark_data(countData = gtexCm,
                                              metaData = txInfo,
                                              filter = "edgeR",
                                              edgeR_filter_spec = list(min.count = 10, 
                                                                       min.total.count = 15, 
                                                                       large.n = 10, 
                                                                       min.prop = 0.7), 
                                              nrRepList = nrRepList, 
                                              fracGenesAffected = 0.15)
names(gtexBenchmarkDataLenient) <- paste0(names(nrRepList),"_filterLenient")

gtexBenchmarkDataStringent <- getBenchmark_data(countData=gtexCm, metaData=txInfo,filter="DRIMSeq", nrRepList=nrRepList, fracGenesAffected=0.15)
names(gtexBenchmarkDataStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save gtex benchmark data

```{r}
#save(gtexBenchmarkDataLenient, gtexBenchmarkDataStringent, file="./Data/GTEx_benchmark_datasets_count.Rdata")
save(gtexBenchmarkDataLenient, gtexBenchmarkDataStringent, file="./Data/GTEx_benchmark_datasets_scaledTPM.Rdata")
```

